### PR TITLE
Fix for burn up/double shock

### LINKED ISSUE #2960: Burn Up doesn't remove Fire-Type if it gets a KO
burnup-bug
Burn Up should remove the user's Fire-type whether it gets a KO or not.

### LINKED ISSUE #2961: Burn Up is affected by the user's Pressure Ability
burnup-bug
Burn Up should only be affected by an opponent's Pressure Ability.

### PR DESCRIPTION
**Pull request by AlexOn1ine**
## Description
Fixes Burn Up and Double Shock not losing the typing when a target is killed and  double pp consumption when used.

/Edit: Added tests including one specific for this scenario that fails without the fix.
## Images
double_shock

## Issue(s) that this PR fixes
Fixes #2961, Fixes #2960

## **Discord contact info**
AlexOnline#2331

### CODE REVIEW

File: test/move_effect_burn_up.c [hunk 1]
```diff
@@ -0,0 +1,57 @@
+#include "global.h"
+#include "test_battle.h"
+
+ASSUMPTIONS
+{
+    ASSUME(gBattleMoves[MOVE_BURN_UP].effect == EFFECT_BURN_UP);
+}
+
+SINGLE_BATTLE_TEST("Burn Up user loses it's typing", s16 damage)
```
Comment (AsparagusEduardo) [inline test/move_effect_burn_up.c:9]:
```suggestion
SINGLE_BATTLE_TEST("Burn Up user loses its Fire typing", s16 damage)
```

File: test/move_effect_burn_up.c [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    } WHEN {
+        TURN { MOVE(player, MOVE_BURN_UP); }
+    } SCENE {
+        NONE_OF { ANIMATION(ANIM_TYPE_MOVE, MOVE_BURN_UP, player); }
+        MESSAGE("Wobbuffet used Burn Up!");
+        MESSAGE("But it failed!");
+    }
+}
+
+SINGLE_BATTLE_TEST("Burn Up user loses it's typing if enemy faints", s16 damage)
```
Comment (AsparagusEduardo) [inline test/move_effect_burn_up.c:42]:
```suggestion
SINGLE_BATTLE_TEST("Burn Up user loses its Fire typing if enemy faints", s16 damage)
```

File: test/move_effect_burn_up.c [hunk 3]
```diff
@@ -0,0 +1,57 @@
+#include "global.h"
+#include "test_battle.h"
+
+ASSUMPTIONS
+{
+    ASSUME(gBattleMoves[MOVE_BURN_UP].effect == EFFECT_BURN_UP);
+}
```
Comment (AsparagusEduardo) [inline test/move_effect_burn_up.c:7]: Needs the assumption that Cyndaquil is a fire type and that Wobbuffet is not.

File: test/move_effect_double_shock.c [hunk 4]
```diff
@@ -0,0 +1,57 @@
+#include "global.h"
+#include "test_battle.h"
+
+ASSUMPTIONS
+{
+    ASSUME(gBattleMoves[MOVE_DOUBLE_SHOCK].effect == EFFECT_DOUBLE_SHOCK);
+}
+
+SINGLE_BATTLE_TEST("Double Shock user loses it's typing", s16 damage)
```
Comment (AsparagusEduardo) [inline test/move_effect_double_shock.c:9]:
```suggestion
SINGLE_BATTLE_TEST("Double Shock user loses its Electric typing", s16 damage)
```

File: test/move_effect_double_shock.c [hunk 5]
```diff
@@ -0,0 +1,57 @@
+#include "global.h"
+#include "test_battle.h"
+
+ASSUMPTIONS
+{
+    ASSUME(gBattleMoves[MOVE_DOUBLE_SHOCK].effect == EFFECT_DOUBLE_SHOCK);
+}
```
Comment (AsparagusEduardo) [inline test/move_effect_double_shock.c:7]: Needs the assumption that Pikachu is an Electric type and that Wobbuffet is not.

File: test/move_effect_double_shock.c [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    } WHEN {
+        TURN { MOVE(player, MOVE_DOUBLE_SHOCK); }
+    } SCENE {
+        NONE_OF { ANIMATION(ANIM_TYPE_MOVE, MOVE_DOUBLE_SHOCK, player); }
+        MESSAGE("Wobbuffet used Double Shock!");
+        MESSAGE("But it failed!");
+    }
+}
+
+SINGLE_BATTLE_TEST("Double Shock user loses it's typing if enemy faints", s16 damage)
```
Comment (AsparagusEduardo) [inline test/move_effect_double_shock.c:42]:
```suggestion
SINGLE_BATTLE_TEST("Double Shock user loses its Electric typing if enemy faints", s16 damage)
```

File: test/move_effect_double_shock.c [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        TURN { MOVE(player, MOVE_DOUBLE_SHOCK); }
+    } SCENE {
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_DOUBLE_SHOCK, player);
+        MESSAGE("Pikachu used up all of its electricity!");
+        MESSAGE("Pikachu used Double Shock!");
+        MESSAGE("But it failed!");
+    }
+}
+
+SINGLE_BATTLE_TEST("Double Shock fails if the user isn't a fire type", s16 damage)
```
Comment (AsparagusEduardo) [inline test/move_effect_double_shock.c:28]:
```suggestion
SINGLE_BATTLE_TEST("Double Shock fails if the user isn't an Electric-type", s16 damage)
```

File: test/move_effect_burn_up.c [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+        TURN { MOVE(player, MOVE_BURN_UP); }
+    } SCENE {
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_BURN_UP, player);
+        MESSAGE("Cyndaquil burned itself out!");
+        MESSAGE("Cyndaquil used Burn Up!");
+        MESSAGE("But it failed!");
+    }
+}
+
+SINGLE_BATTLE_TEST("Burn Up fails if the user isn't a fire type", s16 damage)
```
Comment (AsparagusEduardo) [inline test/move_effect_burn_up.c:28]:
```suggestion
SINGLE_BATTLE_TEST("Burn Up fails if the user isn't a Fire-type", s16 damage)
```

**Review by AsparagusEduardo — CHANGES_REQUESTED** [review]

**Comment by AlexOn1ine** [comment]
Applied all suggestions.

**Review by AsparagusEduardo — APPROVED** [review]

**Pull request merged**